In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import warnings

In [2]:
df = pd.read_csv('../data/DataSet2.csv')
df.head(10)

Unnamed: 0                            Film Oscar Year  \
0           0                           Wings    1927/28   
1           1                      7th Heaven    1927/28   
2           2                      The Racket    1927/28   
3           3             The Broadway Melody    1928/29   
4           4                           Alibi    1928/29   
5           5                 Hollywood Revue    1928/29   
6           6                  In Old Arizona    1928/29   
7           7                     The Patriot    1928/29   
8           8  All Quiet on the Western Front    1929/30   
9           9                   The Big House    1929/30   

  Film Studio/Producer(s)    Award  Year of Release  Movie Time  \
0    Famous Players-Lasky   Winner             1927         144   
1                     Fox  Nominee             1927         110   
2       The Caddo Company  Nominee             1928          84   
3     Metro-Goldwyn-Mayer   Winner             1929         100   
4     Feature Productions  Nominee             1929          91   
5     Metro-Goldwyn-Mayer  Nominee             1929         130   
6                     Fox  Nominee             1928          95   
7  Paramount Famous Lasky  Nominee             1928         113   
8               Universal   Winner             1930         152   
9            Cosmopolitan  Nominee             1930          87   

              Movie Genre  IMDB Rating IMDB Votes  ... Tomatometer Status  \
0       Drama,Romance,War          7.5     12,221  ...    Certified-Fresh   
1           Drama,Romance          7.7      3,439  ...                NaN   
2   Crime,Drama,Film-Noir          6.7      1,257  ...                NaN   
3   Drama,Musical,Romance          5.7      6,890  ...             Rotten   
4    Action,Crime,Romance          5.8        765  ...                NaN   
5            Comedy,Music          5.7      2,004  ...                NaN   
6                 Western          5.6      1,019  ...             Rotten   
7  Drama,History,Thriller          7.4         18  ...                NaN   
8               Drama,War          8.1     59,214  ...                NaN   
9    Crime,Drama,Thriller          7.1      2,079  ...              Fresh   

  Tomatometer Rating Tomatometer Count Audience Status Audience Rating  \
0               93.0              46.0         Upright            78.0   
1                NaN               NaN             NaN             NaN   
2                NaN               NaN             NaN             NaN   
3               33.0              24.0         Spilled            21.0   
4                NaN               NaN             NaN             NaN   
5                NaN               NaN             NaN             NaN   
6               56.0               9.0         Spilled            38.0   
7                NaN               NaN             NaN             NaN   
8                NaN               NaN             NaN             NaN   
9               75.0               8.0         Upright            69.0   

  Audience Count Tomatometer Top Critics Count  \
0         3530.0                           9.0   
1            NaN                           NaN   
2            NaN                           NaN   
3         1813.0                           7.0   
4            NaN                           NaN   
5            NaN                           NaN   
6          356.0                           2.0   
7            NaN                           NaN   
8            NaN                           NaN   
9          323.0                           0.0   

  Tomatometer Fresh Critics Count Tomatometer Rotten Critics Count  \
0                            43.0                              3.0   
1                             NaN                              NaN   
2                             NaN                              NaN   
3                             8.0                             16.0   
4                             NaN       

In [3]:
df.columns

Index(['Unnamed: 0', 'Film', 'Oscar Year', 'Film Studio/Producer(s)', 'Award',
       'Year of Release', 'Movie Time', 'Movie Genre', 'IMDB Rating',
       'IMDB Votes', 'Movie Info', 'Genres', 'Critic Consensus',
       'Content Rating', 'Directors', 'Authors', 'Actors',
       'Original Release Date', 'Streaming Release Date', 'Production Company',
       'Tomatometer Status', 'Tomatometer Rating', 'Tomatometer Count',
       'Audience Status', 'Audience Rating', 'Audience Count',
       'Tomatometer Top Critics Count', 'Tomatometer Fresh Critics Count',
       'Tomatometer Rotten Critics Count', 'Film ID'],
      dtype='object')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 30 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Unnamed: 0                        571 non-null    int64  
 1   Film                              571 non-null    object 
 2   Oscar Year                        571 non-null    object 
 3   Film Studio/Producer(s)           571 non-null    object 
 4   Award                             571 non-null    object 
 5   Year of Release                   571 non-null    int64  
 6   Movie Time                        571 non-null    int64  
 7   Movie Genre                       571 non-null    object 
 8   IMDB Rating                       571 non-null    float64
 9   IMDB Votes                        571 non-null    object 
 10  Movie Info                        438 non-null    object 
 11  Genres                            439 non-null    object 
 12  Critic C

Сразу удалим стобец "Unnamed: 0"

In [5]:
df.drop('Unnamed: 0', axis= 1 , inplace= True )

### Обработка пропущенных знаечний

Посмотрим число пропущенных значений

In [6]:
df.isna().sum()

Film                                  0
Oscar Year                            0
Film Studio/Producer(s)               0
Award                                 0
Year of Release                       0
Movie Time                            0
Movie Genre                           0
IMDB Rating                           0
IMDB Votes                            0
Movie Info                          133
Genres                              132
Critic Consensus                    232
Content Rating                      132
Directors                           134
Authors                             134
Actors                              132
Original Release Date               132
Streaming Release Date              133
Production Company                  132
Tomatometer Status                  132
Tomatometer Rating                  132
Tomatometer Count                   132
Audience Status                     135
Audience Rating                     132
Audience Count                      132


Из данного списка видно, что из 571 фильма, представленного в датасете, у 132 фильмов отсутствует информация в 19 столбцах из 30. Удалим эти фильмы из таблицы, потому что восстановить данные этих столбцов восстановить невозможно, а подставление нулей или медианы очень сильно скажется на оценке содержимого. 

In [7]:
df.dropna(subset=['Production Company'], how='all', inplace=True)

In [8]:
df.isna().sum()

Film                                  0
Oscar Year                            0
Film Studio/Producer(s)               0
Award                                 0
Year of Release                       0
Movie Time                            0
Movie Genre                           0
IMDB Rating                           0
IMDB Votes                            0
Movie Info                            1
Genres                                0
Critic Consensus                    100
Content Rating                        0
Directors                             2
Authors                               2
Actors                                0
Original Release Date                 0
Streaming Release Date                1
Production Company                    0
Tomatometer Status                    0
Tomatometer Rating                    0
Tomatometer Count                     0
Audience Status                       3
Audience Rating                       0
Audience Count                        0


В столбце "Critic Consensus" проставим unknown, потому что восстановить его невозможно, это же сделаем со столбцами "Movie Info", "Directors", "Authors"

In [9]:
df['Critic Consensus'] = df['Critic Consensus'].fillna('Unknown')
df['Movie Info'] = df['Movie Info'].fillna('Unknown')
df['Directors'] = df['Directors'].fillna('Unknown')
df['Authors'] = df['Authors'].fillna('Unknown')

In [10]:
df.isna().sum()

Film                                0
Oscar Year                          0
Film Studio/Producer(s)             0
Award                               0
Year of Release                     0
Movie Time                          0
Movie Genre                         0
IMDB Rating                         0
IMDB Votes                          0
Movie Info                          0
Genres                              0
Critic Consensus                    0
Content Rating                      0
Directors                           0
Authors                             0
Actors                              0
Original Release Date               0
Streaming Release Date              1
Production Company                  0
Tomatometer Status                  0
Tomatometer Rating                  0
Tomatometer Count                   0
Audience Status                     3
Audience Rating                     0
Audience Count                      0
Tomatometer Top Critics Count       0
Tomatometer 

В столбце "Streaming Release Date" 

In [11]:
df_new = df[df['Streaming Release Date'].isnull()]
df_new

Film Oscar Year Film Studio/Producer(s)    Award  \
212  Cat on a Hot Tin Roof       1958     Lawrence Weingarten  Nominee   

     Year of Release  Movie Time Movie Genre  IMDB Rating IMDB Votes  \
212             1958         108       Drama          8.0     47,222   

    Movie Info  ... Tomatometer Status Tomatometer Rating Tomatometer Count  \
212    Unknown  ...              Fresh               97.0              37.0   

    Audience Status Audience Rating Audience Count  \
212         Upright            92.0        31748.0   

    Tomatometer Top Critics Count Tomatometer Fresh Critics Count  \
212                           5.0                            36.0   

    Tomatometer Rotten Critics Count                               Film ID  
212                              1.0  466621aa-287c-4316-9083-74a5fc72a4f6  

[1 rows x 29 columns]

In [12]:
df.loc[df['Film'] == 'Cat on a Hot Tin Roof', 'Production Company']

212    MGM Home Entertainment
Name: Production Company, dtype: object

Посмотрим на дату онлайн релиза фильмов, которые имеют ту же компанию по продакшену

In [13]:
df.loc[df['Production Company'] == 'MGM Home Entertainment', 'Streaming Release Date']

3      2017-02-25
13     2017-01-08
18     2016-12-20
27     2012-04-16
60     2016-11-18
70     2015-12-26
72     2016-11-21
83     2009-01-01
84     2013-12-03
92     2015-09-11
106    2017-03-12
120    2017-01-12
126    2016-11-21
129    2016-05-07
146    2008-08-04
148    2008-12-31
154    2009-04-01
173    2012-04-16
182    2013-12-03
186    2008-01-01
193    2015-05-08
195    2017-01-19
210    2016-04-19
212           NaN
213    2017-04-18
214    2016-05-04
306    2016-07-14
359    2016-11-01
Name: Streaming Release Date, dtype: object

Мы видим, что даты онлайн релизов фильмов, стоящих рядом, находятся в 2016-2017 годах, проставим схлжую дату. 

In [14]:
df.at [212, 'Streaming Release Date'] = '2016-08-10'

Так как столбец "Audience Status" содержит исключительно два значения "Upright" и "Spilled", то проставим вместо NaN самое часто встречающееся значение

In [15]:
df.mode()['Audience Status'][0]

'Upright'

In [16]:
df['Audience Status'] = df['Audience Status'].fillna('Upright')

In [17]:
df.isna().sum()

Film                                0
Oscar Year                          0
Film Studio/Producer(s)             0
Award                               0
Year of Release                     0
Movie Time                          0
Movie Genre                         0
IMDB Rating                         0
IMDB Votes                          0
Movie Info                          0
Genres                              0
Critic Consensus                    0
Content Rating                      0
Directors                           0
Authors                             0
Actors                              0
Original Release Date               0
Streaming Release Date              0
Production Company                  0
Tomatometer Status                  0
Tomatometer Rating                  0
Tomatometer Count                   0
Audience Status                     0
Audience Rating                     0
Audience Count                      0
Tomatometer Top Critics Count       0
Tomatometer 

In [18]:
df

Film Oscar Year  \
0                  Wings    1927/28   
3    The Broadway Melody    1928/29   
6         In Old Arizona    1928/29   
9          The Big House    1929/30   
13              Cimarron    1930/31   
..                   ...        ...   
552       A Star Is Born       2018   
553                 Vice       2018   
557         The Irishman       2019   
559                Joker       2019   
561       Marriage Story       2019   

                               Film Studio/Producer(s)    Award  \
0                                 Famous Players-Lasky   Winner   
3                                  Metro-Goldwyn-Mayer   Winner   
6                                                  Fox  Nominee   
9                                         Cosmopolitan  Nominee   
13                                           RKO Radio   Winner   
..                                                 ...      ...   
552  Bill Gerber, Bradley Cooper, and Lynette Howel...  Nominee   
553  Dede Gardner, Jeremy Kleiner, Adam McKay, and ...  Nominee   
557  Martin Scorsese, Robert De Niro, Jane Rosentha...  Nominee   
559  Todd Phillips, Bradley Cooper, and Emma Tillin...  Nominee   
561                     Noah Baumbach and David Heyman  Nominee   

     Year of Release  Movie Time             Movie Genre  IMDB Rating  \
0               1927         144       Drama,Romance,War          7.5   
3               1929         100   Drama,Musical,Romance          5.7   
6               1928          95                 Western          5.6   
9               1930          87    Crime,Drama,Thriller          7.1   
13              1931         123           Drama,Western          5.9   
..               ...         ...                     ...          ...   
552             2018         136     Drama,Music,Romance          7.6   
553             2018         132  Biography,Comdey,Drama          7.2   
557             2019         209   Biography,Crime,Drama          7.8   
559             2019         122    Crime,Drama,Thriller          8.4   
561             2019         137    Comedy,Drama,Romance          7.9   

    IMDB Votes                                         Movie Info  ...  \
0       12,221  With World War I afoot, David Armstrong (Richa...  ...   
3        6,890  Vaudeville sisters "Hank" (Bessie Love) and Qu...  ...   
6        1,019  In this early Western, notorious bandit the Ci...  ...   
9        2,079  After being convicted of manslaughter, drunk d...  ...   
13       5,851  In 1889, adventurous lawyer and newspaper edit...  ...   
..         ...                                                ...  ...   
552    354,157  Seasoned musician Jackson Maine discovers -- a...  ...   
553    132,014  Gov. George W. Bush of Texas picks Dick Cheney...  ...   
557    349,303  In the 1950s, truck driver Frank Sheeran gets ...  ...   
559  1,056,051  Forever alone in a crowd, failed comedian Arth...  ...   
561    270,304  A stage director and his actor wife struggle t...  ...   

    Tomatometer Status Tomatometer Rating Tomatometer Count Audience Status  \
0      Certified-Fresh               93.0              46.0         Upright   
3               Rotten               33.0              24.0         Spilled   
6               Rotten               56.0               9.0         Spilled   
9                Fresh               75.0               8.0         Upright   
13              Rotten               50.0              24.0         Spilled   
..                 ...                ...               ...             ...   
552    Certified-Fresh               90.0             517.0         Upright   
553              Fresh               65.0             359.0         Upright   
557    Certified-Fresh               96.0             439.0         Upright   
559              Fresh               68.0             574.0         Upright   
561    Certified-Fresh               94.0             381.0         Upright   

    Audience Rating Audience Co

### Проверим уникальность значений

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 439 entries, 0 to 561
Data columns (total 29 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Film                              439 non-null    object 
 1   Oscar Year                        439 non-null    object 
 2   Film Studio/Producer(s)           439 non-null    object 
 3   Award                             439 non-null    object 
 4   Year of Release                   439 non-null    int64  
 5   Movie Time                        439 non-null    int64  
 6   Movie Genre                       439 non-null    object 
 7   IMDB Rating                       439 non-null    float64
 8   IMDB Votes                        439 non-null    object 
 9   Movie Info                        439 non-null    object 
 10  Genres                            439 non-null    object 
 11  Critic Consensus                  439 non-null    object 
 12  Content Ratin

Проверим столбцы "Movie Genre", "Genres", "Content Rating", "Production Company", "Tomatometer Status", "Audience Status".

In [20]:
sorted(df['Movie Genre'].unique())

['Action,Adventure',
 'Action,Adventure,Drama',
 'Action,Adventure,Fantasy',
 'Action,Adventure,History',
 'Action,Adventure,Romance',
 'Action,Adventure,Sci-Fi',
 'Action,Biography,Crime',
 'Action,Crime,Drama',
 'Action,Drama,History',
 'Action,Drama,Thriller',
 'Action,Romance,Thriller',
 'Action,Sci-Fi,Thriller',
 'Adventure,Biography,Crime',
 'Adventure,Biography,Drama',
 'Adventure,Comedy,Crime',
 'Adventure,Comedy,Drama',
 'Adventure,Comedy,Family',
 'Adventure,Comedy,History',
 'Adventure,Comedy,Musical',
 'Adventure,Drama',
 'Adventure,Drama,Family',
 'Adventure,Drama,Fantasy',
 'Adventure,Drama,History',
 'Adventure,Drama,Romance',
 'Adventure,Drama,Sci-Fi',
 'Adventure,Drama,Thriller',
 'Adventure,Drama,War',
 'Adventure,Drama,Western',
 'Adventure,Family,Fantasy',
 'Adventure,Thriller',
 'Animation,Adventure,Comedy',
 'Animation,Family,Fantasy',
 'Biography,Comdey,Drama',
 'Biography,Comedy,Crime',
 'Biography,Comedy,Drama',
 'Biography,Crime,Drama',
 'Biography,Drama',
 'B

In [21]:
sorted(df['Genres'].unique())

['Action & Adventure',
 'Action & Adventure, Animation, Comedy, Kids & Family',
 'Action & Adventure, Art House & International, Classics, Drama',
 'Action & Adventure, Art House & International, Classics, Drama, Romance',
 'Action & Adventure, Classics',
 'Action & Adventure, Classics, Comedy, Kids & Family, Musical & Performing Arts, Science Fiction & Fantasy',
 'Action & Adventure, Classics, Comedy, Romance',
 'Action & Adventure, Classics, Drama',
 'Action & Adventure, Classics, Drama, Mystery & Suspense',
 'Action & Adventure, Classics, Drama, Mystery & Suspense, Romance',
 'Action & Adventure, Classics, Drama, Mystery & Suspense, Science Fiction & Fantasy',
 'Action & Adventure, Classics, Drama, Romance',
 'Action & Adventure, Classics, Drama, Science Fiction & Fantasy',
 'Action & Adventure, Classics, Drama, Western',
 'Action & Adventure, Classics, Romance',
 'Action & Adventure, Classics, Science Fiction & Fantasy',
 'Action & Adventure, Classics, Western',
 'Action & Adventur

In [22]:
sorted(df['Content Rating'].unique())

['G', 'NR', 'PG', 'PG-13', 'R']

In [23]:
sorted(df['Production Company'].unique())

['20th Century Fox',
 '20th Century Fox Film Corporation',
 '20th Century-Fox',
 '21 Laps Entertainment',
 'A24',
 'A24 Films',
 'Annapurna Pictures',
 'Batjac Productions',
 'Buena Vista',
 'Buena Vista Internationa',
 'Buena Vista Pictures',
 'Columbia',
 'Columbia Pictures',
 'Columbia Pictures Corporation',
 'Criterion Collection',
 'DreamWorks Studios',
 'Dreamworks',
 'Dreamworks Distribution',
 'Dreamworks Distribution LLC',
 'Dreamworks Pictures',
 'Eagle-Lion Films',
 'Element Pictures',
 'Embassy Pictures/Rialto Pictures',
 'Film 44',
 'Focus Features',
 'Fox',
 'Fox Searchlight',
 'Fox Searchlight Pictures',
 'Fox Searchlight Pictures, Film4 and Waypoint',
 'Gramercy',
 'Gramercy Pictures',
 'Gravitas',
 'HBO Video',
 'Hollywood/Buena Vista',
 'IFC Films',
 'Image Entertainment Inc.',
 'Liberty Films',
 'Liongate Films',
 'Lions Gate Films',
 'Lionsgate',
 'Live Home Video',
 'MCA Universal Home Video',
 'MGM',
 'MGM Home Entertainment',
 'MGM/United Artists',
 'Madacy Enter

In [24]:
sorted(df['Tomatometer Status'].unique())

['Certified-Fresh', 'Fresh', 'Rotten']

In [25]:
sorted(df['Audience Status'].unique())

['Spilled', 'Upright']

Замены надо производить только в столбце "Production Company"

In [26]:
def dup(df, duplicates, correct):
    for d in duplicates:
        df['Production Company'] = df['Production Company'].replace(d, correct)
    return df

In [27]:
df = dup(df, ['20th Century Fox Film Corporation','20th Century-Fox','Twentieth Century Fox'], '20th Century Fox')
df = dup(df,['A24 Films'],'A24')
df = dup(df,['Buena Vista', 'Hollywood/Buena Vista'],'Buena Vista Pictures')
df = dup(df, ['Columbia Pictures Corporation','Columbia'], 'Columbia Pictures')
df = dup(df, ['DreamWorks Studios', 'Dreamworks'], 'Dreamworks Pictures')
df = dup(df, ['Fox Searchlight', 'Fox Searchlight Pictures, Film4 and Waypoint'], 'Fox Searchlight Pictures')
df = dup(df, ['Gramercy'], 'Gramercy Pictures')
df = dup(df, ['Lions Gate Films', 'Lionsgate'], 'Liongate Films')
df = dup(df, ['Miramax'], 'Miramax Films')
df = dup(df, ['Orion Pictures Corporation'], 'Orion Pictures')
df = dup(df, ['Paramount', 'Paramount Classics', 'Paramount Studios'],'Paramount Pictures')
df = dup(df, ['RKO Radio Pictures'], 'RKO')
df = dup(df, ['Roadside Attractions\xa0'], 'Roadside Attractions')
df = dup(df, ['Sony Classics','Sony Pictures Classics', 'Sony Pictures Entertainment' ], 'Sony Pictures')
df = dup(df, ['The Weinstein Co./MGM', 'The Weinstein Company'], 'The Weinstein Co.')
df = dup(df, ['United Artists Films'], 'United Artists')
df = dup(df, ['Universal','Universal Studios'], 'Universal Pictures')
df = dup(df, ['WARNER BROTHERS PICTURES','WB', 'Warner Bros.'], 'Warner Bros. Pictures')
df = dup(df, ['Walt Disney Productions'], 'Walt Disney Pictures')

### Замена типов столбцов

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 439 entries, 0 to 561
Data columns (total 29 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Film                              439 non-null    object 
 1   Oscar Year                        439 non-null    object 
 2   Film Studio/Producer(s)           439 non-null    object 
 3   Award                             439 non-null    object 
 4   Year of Release                   439 non-null    int64  
 5   Movie Time                        439 non-null    int64  
 6   Movie Genre                       439 non-null    object 
 7   IMDB Rating                       439 non-null    float64
 8   IMDB Votes                        439 non-null    object 
 9   Movie Info                        439 non-null    object 
 10  Genres                            439 non-null    object 
 11  Critic Consensus                  439 non-null    object 
 12  Content Ratin

Удалим все запятые и элементов столбца "IMDB Votes"

In [29]:
df['IMDB Votes'] = df['IMDB Votes'].replace({',':''}, regex=True)

In [30]:
df['Original Release Date'] = pd.to_datetime(df['Original Release Date'],format = '%Y-%m-%d')
df['Streaming Release Date'] = pd.to_datetime(df['Streaming Release Date'],format = '%Y-%m-%d')
df['Year of Release'] = df['Year of Release'].astype(int)
df['Movie Time'] = df['Movie Time'].astype(int)
df['IMDB Votes'] = df['IMDB Votes'].astype(int)
df

Film Oscar Year  \
0                  Wings    1927/28   
3    The Broadway Melody    1928/29   
6         In Old Arizona    1928/29   
9          The Big House    1929/30   
13              Cimarron    1930/31   
..                   ...        ...   
552       A Star Is Born       2018   
553                 Vice       2018   
557         The Irishman       2019   
559                Joker       2019   
561       Marriage Story       2019   

                               Film Studio/Producer(s)    Award  \
0                                 Famous Players-Lasky   Winner   
3                                  Metro-Goldwyn-Mayer   Winner   
6                                                  Fox  Nominee   
9                                         Cosmopolitan  Nominee   
13                                           RKO Radio   Winner   
..                                                 ...      ...   
552  Bill Gerber, Bradley Cooper, and Lynette Howel...  Nominee   
553  Dede Gardner, Jeremy Kleiner, Adam McKay, and ...  Nominee   
557  Martin Scorsese, Robert De Niro, Jane Rosentha...  Nominee   
559  Todd Phillips, Bradley Cooper, and Emma Tillin...  Nominee   
561                     Noah Baumbach and David Heyman  Nominee   

     Year of Release  Movie Time             Movie Genre  IMDB Rating  \
0               1927         144       Drama,Romance,War          7.5   
3               1929         100   Drama,Musical,Romance          5.7   
6               1928          95                 Western          5.6   
9               1930          87    Crime,Drama,Thriller          7.1   
13              1931         123           Drama,Western          5.9   
..               ...         ...                     ...          ...   
552             2018         136     Drama,Music,Romance          7.6   
553             2018         132  Biography,Comdey,Drama          7.2   
557             2019         209   Biography,Crime,Drama          7.8   
559             2019         122    Crime,Drama,Thriller          8.4   
561             2019         137    Comedy,Drama,Romance          7.9   

     IMDB Votes                                         Movie Info  ...  \
0         12221  With World War I afoot, David Armstrong (Richa...  ...   
3          6890  Vaudeville sisters "Hank" (Bessie Love) and Qu...  ...   
6          1019  In this early Western, notorious bandit the Ci...  ...   
9          2079  After being convicted of manslaughter, drunk d...  ...   
13         5851  In 1889, adventurous lawyer and newspaper edit...  ...   
..          ...                                                ...  ...   
552      354157  Seasoned musician Jackson Maine discovers -- a...  ...   
553      132014  Gov. George W. Bush of Texas picks Dick Cheney...  ...   
557      349303  In the 1950s, truck driver Frank Sheeran gets ...  ...   
559     1056051  Forever alone in a crowd, failed comedian Arth...  ...   
561      270304  A stage director and his actor wife struggle t...  ...   

    Tomatometer Status Tomatometer Rating Tomatometer Count Audience Status  \
0      Certified-Fresh               93.0              46.0         Upright   
3               Rotten               33.0              24.0         Spilled   
6               Rotten               56.0               9.0         Spilled   
9                Fresh               75.0               8.0         Upright   
13              Rotten               50.0              24.0         Spilled   
..                 ...                ...               ...             ...   
552    Certified-Fresh               90.0             517.0         Upright   
553              Fresh               65.0             359.0         Upright   
557    Certified-Fresh               96.0             439.0         Upright   
559              Fresh               68.0             574.0         Upright   
561    Certified-Fresh               94.0             381.0         Upright   

    Audience Rating

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 439 entries, 0 to 561
Data columns (total 29 columns):
 #   Column                            Non-Null Count  Dtype         
---  ------                            --------------  -----         
 0   Film                              439 non-null    object        
 1   Oscar Year                        439 non-null    object        
 2   Film Studio/Producer(s)           439 non-null    object        
 3   Award                             439 non-null    object        
 4   Year of Release                   439 non-null    int32         
 5   Movie Time                        439 non-null    int32         
 6   Movie Genre                       439 non-null    object        
 7   IMDB Rating                       439 non-null    float64       
 8   IMDB Votes                        439 non-null    int32         
 9   Movie Info                        439 non-null    object        
 10  Genres                            439 non-null    objec

# Обогащение данных

Столбцы "Original Release Date" и "Streaming Release Date" разберём на дни, месяца и года, но только для "Streaming Release Date", потому что делать такой столбец для "Original Release Date", так как уже существует "Year of Release" - бессмысленно. Смысла брать 

In [32]:
df['Day of Original Release'] = df['Original Release Date'].dt.day_name()
df['Day of Streaming Release'] = df['Streaming Release Date'].dt.day_name()

In [33]:
df['Month of Original Release'] = df['Original Release Date'].dt.month_name()
df['Month of Streaming Release'] = df['Streaming Release Date'].dt.month_name()

In [34]:
df['Year of Streaming Release'] = df['Streaming Release Date'].dt.year

In [35]:
df

Film Oscar Year  \
0                  Wings    1927/28   
3    The Broadway Melody    1928/29   
6         In Old Arizona    1928/29   
9          The Big House    1929/30   
13              Cimarron    1930/31   
..                   ...        ...   
552       A Star Is Born       2018   
553                 Vice       2018   
557         The Irishman       2019   
559                Joker       2019   
561       Marriage Story       2019   

                               Film Studio/Producer(s)    Award  \
0                                 Famous Players-Lasky   Winner   
3                                  Metro-Goldwyn-Mayer   Winner   
6                                                  Fox  Nominee   
9                                         Cosmopolitan  Nominee   
13                                           RKO Radio   Winner   
..                                                 ...      ...   
552  Bill Gerber, Bradley Cooper, and Lynette Howel...  Nominee   
553  Dede Gardner, Jeremy Kleiner, Adam McKay, and ...  Nominee   
557  Martin Scorsese, Robert De Niro, Jane Rosentha...  Nominee   
559  Todd Phillips, Bradley Cooper, and Emma Tillin...  Nominee   
561                     Noah Baumbach and David Heyman  Nominee   

     Year of Release  Movie Time             Movie Genre  IMDB Rating  \
0               1927         144       Drama,Romance,War          7.5   
3               1929         100   Drama,Musical,Romance          5.7   
6               1928          95                 Western          5.6   
9               1930          87    Crime,Drama,Thriller          7.1   
13              1931         123           Drama,Western          5.9   
..               ...         ...                     ...          ...   
552             2018         136     Drama,Music,Romance          7.6   
553             2018         132  Biography,Comdey,Drama          7.2   
557             2019         209   Biography,Crime,Drama          7.8   
559             2019         122    Crime,Drama,Thriller          8.4   
561             2019         137    Comedy,Drama,Romance          7.9   

     IMDB Votes                                         Movie Info  ...  \
0         12221  With World War I afoot, David Armstrong (Richa...  ...   
3          6890  Vaudeville sisters "Hank" (Bessie Love) and Qu...  ...   
6          1019  In this early Western, notorious bandit the Ci...  ...   
9          2079  After being convicted of manslaughter, drunk d...  ...   
13         5851  In 1889, adventurous lawyer and newspaper edit...  ...   
..          ...                                                ...  ...   
552      354157  Seasoned musician Jackson Maine discovers -- a...  ...   
553      132014  Gov. George W. Bush of Texas picks Dick Cheney...  ...   
557      349303  In the 1950s, truck driver Frank Sheeran gets ...  ...   
559     1056051  Forever alone in a crowd, failed comedian Arth...  ...   
561      270304  A stage director and his actor wife struggle t...  ...   

    Audience Count Tomatometer Top Critics Count  \
0           3530.0                           9.0   
3           1813.0                           7.0   
6            356.0                           2.0   
9            323.0                           0.0   
13          1514.0                           6.0   
..             ...                           ...   
552        20442.0                          62.0   
553         5061.0                          47.0   
557         1057.0                          53.0   
559        66303.0                          56.0   
561          184.0                          48.0   

    Tomatometer Fresh Critics Count Tomatometer Rotten Critics Count  \
0                              43.0                              3.0   
3                               8.0                             16.0   
6                               5.0                              4.0   
9                               6.0                    

Добавим столбец, который будет находить разницу между годом оригинального релиза и годом стримингового релиза

In [36]:
df['Difference between Streaming and Original Realease Year'] = df['Year of Streaming Release'] - df['Year of Release']

Добавим столбце, который будет считать среднюю оценку фильма с помощью столбцов "IMDB Rating", "Tomatometer Rating" и "Audience Rating"

In [37]:
df['Avg Score'] = round((df['IMDB Rating']*10 + df['Tomatometer Rating'] + df['Audience Rating'])/3)

Заменим значения в столбцах "Award", "Tomatometer Status", "Audience Status" на числовые

In [38]:
dict_cond = {'Winner': 1,
             'Nominee': 0}
df['Award'] = df['Award'].map(dict_cond)

In [39]:
dict_cond = {'Upright': 1,
             'Spilled': 0}
df['Audience Status'] = df['Audience Status'].map(dict_cond)

In [40]:
dict_cond = {'Certified-Fresh': 2,
             'Fresh' : 1,
             'Rotten': 0}
df['Tomatometer Status'] = df['Tomatometer Status'].map(dict_cond)

In [41]:
df

Film Oscar Year  \
0                  Wings    1927/28   
3    The Broadway Melody    1928/29   
6         In Old Arizona    1928/29   
9          The Big House    1929/30   
13              Cimarron    1930/31   
..                   ...        ...   
552       A Star Is Born       2018   
553                 Vice       2018   
557         The Irishman       2019   
559                Joker       2019   
561       Marriage Story       2019   

                               Film Studio/Producer(s)  Award  \
0                                 Famous Players-Lasky      1   
3                                  Metro-Goldwyn-Mayer      1   
6                                                  Fox      0   
9                                         Cosmopolitan      0   
13                                           RKO Radio      1   
..                                                 ...    ...   
552  Bill Gerber, Bradley Cooper, and Lynette Howel...      0   
553  Dede Gardner, Jeremy Kleiner, Adam McKay, and ...      0   
557  Martin Scorsese, Robert De Niro, Jane Rosentha...      0   
559  Todd Phillips, Bradley Cooper, and Emma Tillin...      0   
561                     Noah Baumbach and David Heyman      0   

     Year of Release  Movie Time             Movie Genre  IMDB Rating  \
0               1927         144       Drama,Romance,War          7.5   
3               1929         100   Drama,Musical,Romance          5.7   
6               1928          95                 Western          5.6   
9               1930          87    Crime,Drama,Thriller          7.1   
13              1931         123           Drama,Western          5.9   
..               ...         ...                     ...          ...   
552             2018         136     Drama,Music,Romance          7.6   
553             2018         132  Biography,Comdey,Drama          7.2   
557             2019         209   Biography,Crime,Drama          7.8   
559             2019         122    Crime,Drama,Thriller          8.4   
561             2019         137    Comedy,Drama,Romance          7.9   

     IMDB Votes                                         Movie Info  ...  \
0         12221  With World War I afoot, David Armstrong (Richa...  ...   
3          6890  Vaudeville sisters "Hank" (Bessie Love) and Qu...  ...   
6          1019  In this early Western, notorious bandit the Ci...  ...   
9          2079  After being convicted of manslaughter, drunk d...  ...   
13         5851  In 1889, adventurous lawyer and newspaper edit...  ...   
..          ...                                                ...  ...   
552      354157  Seasoned musician Jackson Maine discovers -- a...  ...   
553      132014  Gov. George W. Bush of Texas picks Dick Cheney...  ...   
557      349303  In the 1950s, truck driver Frank Sheeran gets ...  ...   
559     1056051  Forever alone in a crowd, failed comedian Arth...  ...   
561      270304  A stage director and his actor wife struggle t...  ...   

    Tomatometer Fresh Critics Count Tomatometer Rotten Critics Count  \
0                              43.0                              3.0   
3                               8.0                             16.0   
6                               5.0                              4.0   
9                               6.0                              2.0   
13                             12.0                             12.0   
..                              ...                              ...   
552                           466.0                             52.0   
553                           234.0                            125.0   
557                           421.0                             19.0   
559                           393.0                            182.0   
561                           359.0                             22.0   

                                  Film ID Day of Original Release  \
0    2becf7d5-a3de-46ab-ae45-abdd6b588067                 

### **Проверка "очищенных" данных**

Проверим строки на дубликаты

In [47]:
df.duplicated().sum()

0

Дубликатов нет. Проверим сводную информацию в DataFrame.

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 439 entries, 0 to 561
Data columns (total 36 columns):
 #   Column                                                   Non-Null Count  Dtype         
---  ------                                                   --------------  -----         
 0   Film                                                     439 non-null    object        
 1   Oscar Year                                               439 non-null    object        
 2   Film Studio/Producer(s)                                  439 non-null    object        
 3   Award                                                    439 non-null    int64         
 4   Year of Release                                          439 non-null    int32         
 5   Movie Time                                               439 non-null    int32         
 6   Movie Genre                                              439 non-null    object        
 7   IMDB Rating                                              4

# EDA